File: 滑动窗口/209.长度最小的子数组.py


# 滑动窗口
class Solution1:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        l, total = 0, 0
        res = float("inf")
        for r in range(len(nums)):  # 右边指针只顾往前走
            total += nums[r]  # 向左边滑动增大值
            while total >= target:  # 满足条件了，我们尝试去找到最小窗口
                res = min(r - l + 1, res)
                total -= nums[l] # 向右边滑动减小值
                l += 1
        return 0 if res == float("inf") else res

File: 滑动窗口/209.长度最小的子数组.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
// NOTE: 滑动窗口问题 - 累计型问题
var minSubArrayLen = function (target, nums) {
  let l = 0, res = target + 1, total = 0;
  for (let r = 0; r < nums.length; r++) {
    total += nums[r]; //! r = 0 时窗口内应该存在一个值
    while (total >= target) {
      //! 等于的时候也需要调整
      res = Math.min(res, r - l + 1);
      total -= nums[l];
      l++;
    }
  }
  return res > target ? 0 : res;
};
